### PR TITLE
[Model monitoring] Save current_stats, drift_measures to json file

### PR DESCRIPTION
**Pull request by royischoss**
https://iguazio.atlassian.net/browse/ML-7461
Saving current stats and drift measures of histogram_data_drift application as a json file.

### CODE REVIEW

File: mlrun/model_monitoring/applications/histogram_data_drift.py [hunk 1]
```diff
@@ -214,6 +217,30 @@ def _get_metrics(
 
         return metrics
 
+    @staticmethod
+    def _get_stats(
+        metrics: list[mm_results.ModelMonitoringApplicationMetric],
+        metrics_per_feature: DataFrame,
+        monitoring_context: mm_context.MonitoringApplicationContext,
+    ):
```
Comment (davesh0812) [inline mlrun/model_monitoring/applications/histogram_data_drift.py]: please add return type hint & short explain

File: mlrun/model_monitoring/applications/histogram_data_drift.py [hunk 2]
```diff
@@ -336,9 +363,13 @@ def do_tracking(
...
             metrics_per_feature=metrics_per_feature,
         )
-
-        metrics_and_result = metrics + [result]
+        stats = self._get_stats(
+            metrics=metrics,
+            monitoring_context=monitoring_context,
+            metrics_per_feature=metrics_per_feature,
+        )
+        metrics_result_and_stats = metrics + [result] + stats
```
Comment (davesh0812) [inline mlrun/model_monitoring/applications/histogram_data_drift.py:376]: please edit the the return type according to this change

File: mlrun/model_monitoring/applications/results.py [hunk 3]
```diff
@@ -121,3 +122,28 @@ def to_dict(self):
             mm_constant.MetricData.METRIC_NAME: self.name,
             mm_constant.MetricData.METRIC_VALUE: self.value,
         }
+
+
+@dataclasses.dataclass
+class _ModelMonitoringApplicationStats(_ModelMonitoringApplicationDataRes):
+    """
+    Class representing the stats of a custom model monitoring application.
```
Comment (davesh0812) [inline mlrun/model_monitoring/applications/results.py]: used only for histogram app, can you edit this line?

File: tests/system/model_monitoring/test_app.py [hunk 4]
```diff
@@ -765,8 +777,8 @@ def test_app_flow(self, with_training_set: bool) -> None:
         self._test_artifacts(ep_id=ep_id)
         self._test_api(ep_id=ep_id)
         # TODO: uncomment the following 2 lines once the new API for getting model endpoint stats is implemented
```
Comment (davesh0812) [inline tests/system/model_monitoring/test_app.py]: can delete the todo

File: mlrun/model_monitoring/db/_stats.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    When working with the schedules data, use this class as a context manager to read and write the data.
+    """
+
+    def __init__(self, item: mlrun.datastore.base.DataItem, file_type: str):
+        self._path = item.url
+        self._item = item
+        self._file_type = file_type
+        self._fs = cast(fsspec.AbstractFileSystem, self._item.store.filesystem)
+
+    def create(self, data: Optional[dict] = None) -> None:
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/_stats.py]: create only with empty dict.
Add `write(stats:dict or str, time: datetime or time)`

File: mlrun/model_monitoring/db/_stats.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+from mlrun.errors import MLRunInvalidArgumentTypeError
+from mlrun.model_monitoring.helpers import (
+    get_monitoring_current_stats_data,
+    get_monitoring_drift_measures_data,
+    get_monitoring_stats_directory_path,
+)
+from mlrun.utils import logger
+
+
+class ModelMonitoringStatsFile(AbstractContextManager):
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/_stats.py]: remove `AbstractContextManager` and the with statement

File: tests/system/model_monitoring/test_app.py [hunk 7]
```diff
@@ -708,7 +720,7 @@ def _test_error_alert(cls) -> None:
         # Validate alert notification
         assert alert.count == 1
 
-    @pytest.mark.parametrize("with_training_set", [True, False])
+    @pytest.mark.parametrize("with_training_set", [True])
```
Comment (davesh0812) [inline tests/system/model_monitoring/test_app.py]: please revert

**Review by davesh0812 — COMMENTED** [review]
Nice, few comments.

File: mlrun/model_monitoring/applications/results.py [hunk 8]
```diff
@@ -121,3 +122,31 @@ def to_dict(self):
...
+    """
+    Class representing the stats of histogram data drift application.
+
+    :param name             (str) Enum mm_constant.StatsData str representation of the stats data kind of the event
+    :param                  (str) iso format representation of the timestamp the event took place
+    :param stats            (dict) Dictionary representation of the stats calculated for the event
+
+    """
+
+    name: typing.Union[mm_constant.StatsData, str]
```
Comment (davesh0812) [inline mlrun/model_monitoring/applications/results.py]:
```suggestion
    name: mm_constant.StatsKind
```

File: mlrun/model_monitoring/db/_stats.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+from mlrun.common.schemas.model_monitoring.constants import StatsKind
+from mlrun.model_monitoring.helpers import (
+    get_monitoring_current_stats_data,
+    get_monitoring_drift_measures_data,
+    get_monitoring_stats_directory_path,
+)
+from mlrun.utils import logger
+
+
+class ModelMonitoringStatsFile:
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/_stats.py]: Is it an abstract class?
```suggestion
class ModelMonitoringStatsFile(abc.ABC):
```

and add if possible an abstract method  named `from_model_endpoint()`

File: mlrun/model_monitoring/db/_stats.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+            if timestamp is not None:
+                timestamp = datetime.fromisoformat(timestamp).astimezone(
+                    tz=timezone.utc
+                )
+            return json_dictionary.get("data"), timestamp
+        except Exception as e:
+            logger.debug(
+                f"Model Monitoring {self._file_type} file read error: {e} file might not been created",
+                path=self._item.url,
+            )
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/_stats.py]:
```suggestion
            logger.debug(
                f"Model Monitoring {self._file_type} file read error.",
                path=self._item.url,
                error=e
            )
```

File: mlrun/model_monitoring/db/_stats.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+            raise e
+
+    def write(self, stats: dict, timestamp: datetime) -> None:
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/_stats.py:110]: add some explain
```suggestion
    def write(self, stats: dict, timestamp: datetime) -> None:
    """ overwrite the existing file """
```

File: mlrun/model_monitoring/db/tsdb/tdengine/tdengine_connector.py [hunk 12]
```diff
@@ -531,14 +531,14 @@ def get_last_request(
             start=start,
             end=end,
             columns=[
-                mm_schemas.SchedulingKeys.ENDPOINT_ID,
+                mm_schemas.EventFieldType.ENDPOINT_ID,
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/tsdb/tdengine/tdengine_connector.py:534]: Nice Catch!!

File: mlrun/model_monitoring/db/_stats.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        """
+        try:
+            json_dictionary = json.loads(self._item.get().decode())
+            timestamp = json_dictionary.get("timestamp")
+            if timestamp is not None:
+                timestamp = datetime.fromisoformat(timestamp).astimezone(
+                    tz=timezone.utc
+                )
+            return json_dictionary.get("data"), timestamp
+        except Exception as e:
```
Comment (davesh0812) [inline mlrun/model_monitoring/db/_stats.py]: ?

File: mlrun/model_monitoring/writer.py [hunk 14]
```diff
@@ -197,14 +207,54 @@ def _reconstruct_event(event: _RawEvent) -> tuple[_AppResultEvent, WriterEventKi
...
+            )
+        elif stat_kind == StatsKind.DRIFT_MEASURES.value:
+            ModelMonitoringDriftMeasuresFile(self.project, endpoint_id).write(
+                data, timestamp
+            )
+            logger.info(
+                "Updating the model endpoint with metadata specific to the histogram "
+                "data drift app Drift measures",
+                endpoint_id=endpoint_id,
+            )
```
Comment (davesh0812) [inline mlrun/model_monitoring/writer.py]:
```suggestion
        if stat_kind == StatsKind.CURRENT_STATS:
            ModelMonitoringCurrentStatsFile(self.project, endpoint_id).write(
                data, timestamp
            )
        elif stat_kind == StatsKind.DRIFT_MEASURES:
            ModelMonitoringDriftMeasuresFile(self.project, endpoint_id).write(
                data, timestamp
            )
        logger.info(
            "Updating the model endpoint statistics"
            endpoint_id=endpoint_id,
            stats_kind=stat_kind
        )
```

**Review by davesh0812 — COMMENTED** [review]
few last comments

**Review by davesh0812 — APPROVED** [review]

**Comment by royischoss** [comment]
Passed before mrege - test_app_flow waiting for requirement change  [nuclio-jupyter~=0.10.4]
Changed in [PR](https://github.com/mlrun/mlrun/pull/6645)

**Review by assaf758 — APPROVED** [review]

**Pull request merged**